pkg/engine/lookup_func.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright The Helm Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package engine
+
+import (
+	"log"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+)
+
+type lookupFunc = func(apiversion string, resource string, namespace string, name string) (map[string]interface{}, error)
+
+func NewLookupFunction(config *rest.Config) lookupFunc {
+	return func(apiversion string, resource string, namespace string, name string) (map[string]interface{}, error) {
+		var client dynamic.ResourceInterface
+		c, namespaced, err := getDynamicClientOnKind(apiversion, resource, config)
+		if err != nil {
+			return map[string]interface{}{}, err
+		}
+		if namespaced && namespace != "" {
```

Comment:
What should be the behavior if `namespaced` is true, but the `namespace` is an empty string? Will there be any problems when it uses the non-namespaced client?

Reply:
Nit: Also, consider just passing the namespace into `getDynamicClientOnKind` and have it handle this logic. This logic seems to be a bit out of place here

Reply:
maybe we should add this to the doc, this is the intended behavior:

- namespaced true + non-empty namespace + non empty name -> kubectl get xxx <name> -n <namespace> -> single object
- namespaced true + non-empty namespace + empty name -> kubectl get xxx -n <namespace> -> list
- namespaced true + empty namespace (name is ignored) -> kubectl get xxx --all-namespaces -> list
- namespaced false (namespace is ignored) + non-empty name -> kubectl get xxx <name> -> single object
- namespaced false (namespace is ignored) + empty name -> kubectl get xxx -> list

Please help me make sure the algorithm honors the expected behavior.
This should also explain why we need this logic here and not in the `getDynamicClientOnKind`: we need to handle non-namespaced resources.

Reply:
Yep, that makes sense. Thanks for the explanation. We should definitely add this to the documentation when this get merged